exercises/bracket-push/example.rs

```diff
@@ -13,46 +11,65 @@ impl<'a> From<&'a str> for Brackets {
     }
 }
 
+#[derive(Clone)]
+pub struct MatchingBrackets {
+    collection: HashMap<char, char>,
+}
+
+impl From<Vec<(char, char)>> for MatchingBrackets {
+    fn from(i: Vec<(char, char)>) -> Self {
+        let mut c: HashMap<char, char> = HashMap::new();
+
+        for (k, v) in i {
+            c.insert(k, v);
+        }
+
+        MatchingBrackets { collection: c }
+    }
+}
+
+impl MatchingBrackets {
+    fn contains(&self, other: &char) -> bool {
+        let known = self.collection.keys().chain(self.collection.values()).collect::<Vec<_>>();
+        known.contains(&other)
+    }
+
+    fn closer_for(&self, k: &char) -> Option<&char> {
+        self.collection.get(k)
+    }
+}
+
 impl Brackets {
-    pub fn new(s: String, pairs: Option<HashMap<char, char>>) -> Self {
+    pub fn new(s: String, pairs: Option<Vec<(char, char)>>) -> Self {
         let p = match pairs {
-            Some(x) => x,
-            None => Brackets::default(),
+            Some(x) => MatchingBrackets::from(x),
+            None => MatchingBrackets::from(vec![('[', ']'), ('{', '}'), ('(', ')')]),
         };
 
-        let valid = p.keys().chain(p.values()).collect::<Vec<_>>();
-        let all = s.chars().filter(|c| valid.contains(&c)).collect::<Vec<char>>();
-
         Brackets {
-            all: all,
+            raw_brackets: s.chars().filter(|c| p.contains(&c)).collect::<Vec<char>>(),
             pairs: p.clone(),
         }
     }
 
-    fn default() -> HashMap<char, char> {
-        let mut d = HashMap::new();
-        d.insert('[', ']');
-        d.insert('{', '}');
-        d.insert('(', ')');
-        d
-    }
-
     pub fn are_balanced(&self) -> bool {
         let mut unmatched: Vec<char> = Vec::new();
 
-        for b in self.all.clone() {
+        for &bracket in self.raw_brackets.iter() {
             if unmatched.is_empty() {
-                unmatched.push(b);
+                unmatched.push(bracket);
             } else {
-                let t = unmatched.pop().unwrap();
-
-                if let Some(c) = self.pairs.get(&t) {
-                    if *c != b {
-                        unmatched.push(t);
-                        unmatched.push(b);
+                if let Some(last_unmatched) = unmatched.pop() {
```

Comment:
I expect that will always be `Some`, because if you get here, then `unmatched.is_empty()` is false. Since the `else` on this `if let` is the same as what happens if `unmatched` is empty, the two can be combined and deduplicated.